src/state-managers/url-state-manager.js

```diff
@@ -1,24 +1,50 @@
 // @flow
 
+function setSearchParam(key: string, value: string) {
+  const searchParams = new URLSearchParams(
+    new URL(window.location.href).searchParams,
+  );
+  searchParams.set(key, value);
+  return searchParams.toString() || '';
+}
+
+function getSearchParam(key: string) {
+  const { href } = window.location;
+  const searchParams = new URLSearchParams(new URL(href).searchParams);
+  return searchParams.get(key) || '';
+}
+
 export const UrlStateManager: Losen$StateManager = {
   updateStep: (currentStepName, nextStepName) => {
     const currentUrl = window.location.href;
     const basePath = currentUrl.includes('?')
       ? currentUrl.split('?')[0]
       : currentUrl;
-    const searchParams = new URLSearchParams(new URL(currentUrl).searchParams);
-    searchParams.set('step', nextStepName);
+    const searchParams = setSearchParam('step', nextStepName);
 
     window.history.pushState({ activeStep: currentStepName }, '', currentUrl);
     window.history.replaceState(
       { activeStep: nextStepName },
       '',
-      `${basePath}?${searchParams.toString()}`,
+      `${basePath}?${searchParams}`,
     );
   },
   getActiveStep: () => {
-    const { href } = window.location;
-    const searchParams = new URLSearchParams(new URL(href).searchParams);
-    return searchParams.get('step');
+    return getSearchParam('step');
+  },
+  getItem: (key: string) => {
+    return getSearchParam(key);
+  },
+  setItem: (key: string, value: string) => {
+    const searchParams = setSearchParam(key, value);
+    const currentUrl = window.location.href;
+    const basePath = currentUrl.includes('?')
+      ? currentUrl.split('?')[0]
+      : currentUrl;
+    window.history.replaceState(
+      { ...window.history.state, [key]: value },
```

Comment:
How will the back button work if the user selects multiple options on the page? Will "back" take you back to the previous step, or the previous state?

Reply:
Nice spot!
"replaceState" only replaces the contents of the url. It does not add the page to the history stack.
"pushState" adds the page to the history stack.
This means that "back" will take you to the previous step that was pushed to the history stack.